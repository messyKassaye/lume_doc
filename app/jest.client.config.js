// eslint-disable-next-line import/no-extraneous-dependencies
const { defaults } = require('jest-config');

module.exports = {
  name: 'client',
  displayName: 'Client',
  testMatch: ['**/react/**/specs/*spec.(j|t)s?(x)'],
  testPathIgnorePatterns: [],
  testEnvironment: 'node',
  setupFilesAfterEnv: ['<rootDir>/setUpJestClient.js'],
  moduleFileExtensions: [...defaults.moduleFileExtensions, 'd.ts'],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
    '^shared/(.*)': '<rootDir>/shared/$1',
    '^app/(.*)': '<rootDir>/react/$1',
    '^app/UI/(.*)': '<rootDir>/react/UI/$1',
  },
  snapshotSerializers: ['enzyme-to-json/serializer'],
};
