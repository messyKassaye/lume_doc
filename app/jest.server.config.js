// eslint-disable-next-line import/no-extraneous-dependencies
const { defaults } = require('jest-config');

module.exports = {
  name: 'server',
  displayName: 'Server',
  testMatch: ['**/api/**/specs/*spec.(j|t)s?(x)', '**/shared/**/specs/*spec.(j|t)s?(x)'],
  testEnvironment: 'node',
  setupFilesAfterEnv: ['<rootDir>/setUpJestServer.js'],
  moduleFileExtensions: [...defaults.moduleFileExtensions, 'd.ts'],
  moduleNameMapper: {
    '^api/(.*)': '<rootDir>/api/$1',
    '^shared/(.*)': '<rootDir>/shared/$1',
  },
};
