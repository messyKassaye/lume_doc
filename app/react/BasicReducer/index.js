import createReducer, { actions } from 'app/BasicReducer/reducer';

export default createReducer;
export { actions };
