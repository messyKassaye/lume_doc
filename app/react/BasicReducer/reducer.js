/** @format */

import Immutable from 'immutable';

const SET = 'SET';
const SET_IN = 'SET_IN';
const UPDATE = 'UPDATE';
const UPDATE_IN = 'UPDATE_IN';
const UNSET = 'UNSET';
const REMOVE = 'REMOVE';
const PUSH = 'PUSH';
const CONCAT = 'CONCAT';
const CONCAT_IN = 'CONCAT_IN';

export default function createReducer(namespace, defaultValue) {
  return (currentState = defaultValue, action = {}) => {
    let index;

    switch (action.type) {
      case `${namespace}/${SET}`:
        return Immutable.fromJS(action.value);

      case `${namespace}/${SET_IN}`:
        return currentState.set(action.key, Immutable.fromJS(action.value));

      case `${namespace}/${UNSET}`:
        return Immutable.fromJS(defaultValue);

      case `${namespace}/${PUSH}`:
        return currentState.push(Immutable.fromJS(action.value));

      case `${namespace}/${CONCAT}`:
        return currentState.concat(Immutable.fromJS(action.value));

      case `${namespace}/${CONCAT_IN}`:
        return currentState.updateIn(action.key, collection =>
          collection.concat(Immutable.fromJS(action.value))
        );

      case `${namespace}/${REMOVE}`:
        return Immutable.fromJS(currentState).filter(
          object => object.get('_id') !== action.value._id
        );

      case `${namespace}/${UPDATE}`:
        if (currentState instanceof Immutable.Map) {
          return currentState.merge(action.value);
        }

        index = currentState.findIndex(o => o.get('_id') === action.value._id);
        if (index === -1) {
          return currentState.push(Immutable.fromJS(action.value));
        }
        return currentState.set(index, Immutable.fromJS(action.value));

      case `${namespace}/${UPDATE_IN}`:
        index = currentState.getIn(action.key).findIndex(o => o.get('_id') === action.value._id);
        if (index === -1) {
          return currentState.updateIn(action.key, collection =>
            collection.push(Immutable.fromJS(action.value))
          );
        }
        return currentState.setIn([...action.key, index], Immutable.fromJS(action.value));

      default:
        return Immutable.fromJS(currentState);
    }
  };
}

export function update(namespace, value) {
  return {
    type: `${namespace}/${UPDATE}`,
    value,
  };
}

export function updateIn(namespace, key, value) {
  return {
    type: `${namespace}/${UPDATE_IN}`,
    key,
    value,
  };
}

export function set(namespace, value) {
  return {
    type: `${namespace}/${SET}`,
    value,
  };
}

export function setIn(namespace, key, value) {
  return {
    type: `${namespace}/${SET_IN}`,
    key,
    value,
  };
}

export function unset(namespace) {
  return {
    type: `${namespace}/${UNSET}`,
  };
}

export function push(namespace, value) {
  return {
    type: `${namespace}/${PUSH}`,
    value,
  };
}

export function concat(namespace, value) {
  return {
    type: `${namespace}/${CONCAT}`,
    value,
  };
}

export function concatIn(namespace, key, value) {
  return {
    type: `${namespace}/${CONCAT_IN}`,
    key,
    value,
  };
}

export function remove(namespace, value) {
  return {
    type: `${namespace}/${REMOVE}`,
    value,
  };
}

export const actions = {
  update,
  updateIn,
  set,
  setIn,
  unset,
  push,
  concat,
  concatIn,
  remove,
};
