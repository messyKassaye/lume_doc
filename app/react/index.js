import React from 'react';
import ReactDOM from 'react-dom';

import { AppContainer } from 'react-hot-loader';

import App from './App';

import './App/sockets';

const render = Component => {
  ReactDOM.hydrate(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('root')
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => {
    const nextApp = require('./App');
    render(nextApp);
  });
}
