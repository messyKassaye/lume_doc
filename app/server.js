/* eslint-disable no-console */

import bodyParser from 'body-parser';
import compression from 'compression';
import express from 'express';
import helmet from 'helmet';
import { Server } from 'http';
import mongoose from 'mongoose';
import path from 'path';

import { TaskProvider } from 'shared/tasks/tasks';

import { appContextMiddleware } from 'api/utils/appContextMiddleware';
import uwaziMessage from '../message';
import apiRoutes from './api/api';
import privateInstanceMiddleware from './api/auth/privateInstanceMiddleware';
import authRoutes from './api/auth/routes';
import { config } from './api/config';
import vaultSync from './api/evidences_vault';

import { migrator } from './api/migrations/migrator';
import settings from './api/settings';
import syncWorker from './api/sync/syncWorker';
import errorHandlingMiddleware from './api/utils/error_handling_middleware';
import handleError from './api/utils/handleError.js';
import repeater from './api/utils/Repeater';
import serverRenderingRoutes from './react/server.js';
import { DB } from './api/odm';
import { tenants } from './api/tenants/tenantContext';
import { multitenantMiddleware } from './api/utils/multitenantMiddleware';
import { staticFilesMiddleware } from './api/utils/staticFilesMiddleware';
import { customUploadsPath, uploadsPath } from './api/files/filesystem';
import { tocService } from './api/toc_generation/tocService';

mongoose.Promise = Promise;

const app = express();
app.use(helmet());

const http = Server(app);

const uncaughtError = error => {
  handleError(error, { uncaught: true });
  throw error;
};

process.on('unhandledRejection', uncaughtError);
process.on('uncaughtException', uncaughtError);

const oneYear = 31557600;

let maxage = 0;
if (app.get('env') === 'production') {
  maxage = oneYear;
}

app.use(compression());
app.use(express.static(path.resolve(__dirname, '../dist'), { maxage }));
app.use('/public', express.static(config.publicAssets));
app.use(/\/((?!remotepublic).)*/, bodyParser.json({ limit: '1mb' }));

app.use(appContextMiddleware);

//////
// this middleware should go just before any other that accesses to db
app.use(multitenantMiddleware);
//////

let dbAuth = {};

if (process.env.DBUSER) {
  dbAuth = {
    auth: { authSource: 'admin' },
    user: process.env.DBUSER,
    pass: process.env.DBPASS,
  };
}

console.info('==> Connecting to', config.DBHOST);
DB.connect(config.DBHOST, dbAuth).then(async () => {
  await tenants.setupTenants();
  authRoutes(app);
  app.use(privateInstanceMiddleware);
  app.use('/flag-images', express.static(path.resolve(__dirname, '../dist/flags')));
  app.use('/assets/:fileName', staticFilesMiddleware([customUploadsPath]));
  app.use(
    '/uploaded_documents/:fileName',
    staticFilesMiddleware([
      uploadsPath,
      // retained for backwards compatibility
      customUploadsPath,
      //
    ])
  );
  apiRoutes(app, http);
  serverRenderingRoutes(app);
  app.use(errorHandlingMiddleware);

  if (!config.multiTenant && !config.clusterMode) {
    await tenants.run(async () => {
      const shouldMigrate = await migrator.shouldMigrate();
      if (shouldMigrate) {
        console.error(
          '\x1b[33m%s\x1b[0m',
          '==> Your database needs to be migrated, please run:\n\n yarn migrate & yarn reindex\n\n'
        );
        process.exit();
      }
    });
  }

  const bindAddress = { true: 'localhost' }[process.env.LOCALHOST_ONLY];
  const port = config.PORT;

  http.listen(port, bindAddress, async () => {
    await tenants.run(async () => {
      if (!config.multiTenant && !config.clusterMode) {
        syncWorker.start();

        const { evidencesVault, features } = await settings.get();
        if (evidencesVault && evidencesVault.token && evidencesVault.template) {
          console.info('==> 📥  evidences vault config detected, started sync ....');
          repeater.start(
            () => vaultSync.sync(evidencesVault.token, evidencesVault.template),
            10000
          );
        }

        if (features && features.tocGeneration && features.tocGeneration.url) {
          console.info('==> 🗂️ automatically generating TOCs using external service');
          const service = tocService(features.tocGeneration.url);
          repeater.start(() => service.processNext(), 10000);
        }

        repeater.start(
          () =>
            TaskProvider.runAndWait('TopicClassificationSync', 'TopicClassificationSync', {
              mode: 'onlynew',
              noDryRun: true,
              overwrite: true,
            }),
          10000
        );
      }
    });

    console.info(
      '==> 🌎 Listening on port %s. Open up http://localhost:%s/ in your browser.',
      port,
      port
    );

    if (process.env.HOT) {
      console.info('');
      console.info('==> 📦 webpack is watching...');
      console.info(uwaziMessage);
    }
  });
});
