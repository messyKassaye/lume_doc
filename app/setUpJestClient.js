/* eslint-disable import/no-extraneous-dependencies */
const { configure } = require('enzyme');
const Adapter = require('enzyme-adapter-react-16');

configure({ adapter: new Adapter() });

const warn = console.warn.bind(console);
console.warn = function(message) {
  if (message.match('UNSAFE_')) {
    return;
  }
  warn(message);
};

const error = console.error.bind(console);
console.error = function(message) {
  if (message.match('/api/i18n/systemKeys')) {
    return;
  }
  error(message);
};

process.env.__testingEnvironment = true;

process.on('unhandledRejection', err => {
  fail(err);
});

jasmine.createSpyObj = (name, methodNames) => {
  let names = methodNames;
  if (Array.isArray(name)) {
    names = name;
  }

  const obj = {};

  for (let i = 0; i < names.length; i += 1) {
    obj[names[i]] = jasmine.createSpy(names[i]);
  }

  return obj;
};

const clock = {
  install: jest.useFakeTimers,
  uninstall: jest.clearAllTimers,
  tick: jest.advanceTimersByTime,
};
jasmine.clock = () => clock;
