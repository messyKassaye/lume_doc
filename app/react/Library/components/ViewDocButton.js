/** @format */

import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Map } from 'immutable';
import { t, I18NLink } from 'app/I18N';
import { Icon } from 'UI';
import { actions } from 'app/BasicReducer';
import url from 'url';

function getDocumentUrlQuery(searchTerm, targetReference) {
  const query = {};
  if (searchTerm) {
    query.searchTerm = searchTerm;
  }
  if (targetReference) {
    query.ref = targetReference.get('_id');
  }
  return query;
}

export class ViewDocButton extends Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  onClick(e) {
    e.stopPropagation();
    const { targetReference, openReferencesTab } = this.props;
    if (targetReference) {
      openReferencesTab();
    }
  }

  render() {
    const { sharedId, processed, searchTerm, file, targetReference } = this.props;
    const isEntity = !file;

    const pathname = `/entity/${sharedId}`;
    const query = getDocumentUrlQuery(searchTerm, targetReference);
    const documentViewUrl = url.format({
      pathname,
      query,
    });

    if (!processed && !isEntity) {
      return false;
    }

    return (
      <I18NLink
        to={documentViewUrl}
        className="btn btn-default btn-xs view-doc"
        onClick={this.onClick}
      >
        <Icon icon="angle-right" directionAware /> {t('System', 'View')}
      </I18NLink>
    );
  }
}

ViewDocButton.defaultProps = {
  searchTerm: '',
  processed: false,
  targetReference: null,
};

ViewDocButton.propTypes = {
  file: PropTypes.object,
  sharedId: PropTypes.string.isRequired,
  processed: PropTypes.bool,
  searchTerm: PropTypes.string,
  targetReference: PropTypes.instanceOf(Map),
  openReferencesTab: PropTypes.func.isRequired,
};

export function mapStateToProps(state, props) {
  return {
    searchTerm: props.storeKey ? state[props.storeKey].search.searchTerm : '',
  };
}

export function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      openReferencesTab: () => _dispatch =>
        _dispatch(actions.set('viewer.sidepanel.tab', 'references')),
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(ViewDocButton);
