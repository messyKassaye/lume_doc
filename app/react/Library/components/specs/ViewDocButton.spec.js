import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';

import { actions } from 'app/BasicReducer';

import { ViewDocButton, mapDispatchToProps } from '../ViewDocButton';

describe('ViewDocButton', () => {
  let props;
  let component;
  beforeEach(() => {
    props = {
      type: 'entity',
      format: 'format',
      sharedId: '123',
      searchTerm: '',
      openReferencesTab: jest.fn(),
    };
  });

  const render = () => {
    component = shallow(<ViewDocButton {...props} />);
    return component;
  };

  it('should render a view button poiting to the doc url with the searchTerm if pressent', () => {
    render();
    expect(component).toMatchSnapshot();

    props.searchTerm = 'something';
    render();
    expect(component).toMatchSnapshot();
  });
  describe('when targetReference is provided', () => {
    it('should render view button with reference id in the url query', () => {
      props.targetReference = fromJS({ _id: 'ref1', range: { start: 200, end: 300 } });
      render();
      expect(component).toMatchSnapshot();

      props.searchTerm = 'something';
      render();
      expect(component).toMatchSnapshot();
    });
    it('should call openReferencesTab when clicked', () => {
      const event = { stopPropagation: jest.fn() };
      props.targetReference = fromJS({ range: { start: 200, end: 300 } });
      render();
      component.simulate('click', event);
      expect(props.openReferencesTab).toHaveBeenCalled();
    });
  });

  describe('mapDispatchToProps', () => {
    describe('openReferencesTab', () => {
      it('should set the sidepanel tab to references', () => {
        const innerDispatch = jest.fn();
        const dispatch = jest.fn(fn => fn(innerDispatch));
        const mappedProps = mapDispatchToProps(dispatch);
        mappedProps.openReferencesTab();
        expect(innerDispatch).toHaveBeenCalledWith(
          actions.set('viewer.sidepanel.tab', 'references')
        );
      });
    });
  });
});
