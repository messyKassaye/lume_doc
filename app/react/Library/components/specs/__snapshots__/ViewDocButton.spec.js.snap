// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewDocButton should render a view button poiting to the doc url with the searchTerm if pressent 1`] = `
<Connect(I18NLink)
  className="btn btn-default btn-xs view-doc"
  onClick={[Function]}
  to="/entity/123"
>
  <Connect(Icon)
    directionAware={true}
    icon="angle-right"
  />
   
  View
</Connect(I18NLink)>
`;

exports[`ViewDocButton should render a view button poiting to the doc url with the searchTerm if pressent 2`] = `
<Connect(I18NLink)
  className="btn btn-default btn-xs view-doc"
  onClick={[Function]}
  to="/entity/123?searchTerm=something"
>
  <Connect(Icon)
    directionAware={true}
    icon="angle-right"
  />
   
  View
</Connect(I18NLink)>
`;

exports[`ViewDocButton when targetReference is provided should render view button with reference id in the url query 1`] = `
<Connect(I18NLink)
  className="btn btn-default btn-xs view-doc"
  onClick={[Function]}
  to="/entity/123?ref=ref1"
>
  <Connect(Icon)
    directionAware={true}
    icon="angle-right"
  />
   
  View
</Connect(I18NLink)>
`;

exports[`ViewDocButton when targetReference is provided should render view button with reference id in the url query 2`] = `
<Connect(I18NLink)
  className="btn btn-default btn-xs view-doc"
  onClick={[Function]}
  to="/entity/123?searchTerm=something&ref=ref1"
>
  <Connect(Icon)
    directionAware={true}
    icon="angle-right"
  />
   
  View
</Connect(I18NLink)>
`;
